tests/parser/syntax/test_msg_data.py

```diff
@@ -0,0 +1,57 @@
+import pytest
+from eth_tester.exceptions import TransactionFailed
+
+from vyper import compiler
+from vyper.exceptions import StateAccessViolation
+
+
+def test_msg_data_assignment(get_contract):
+    code = """
+@external
+def foo() -> Bytes[4]:
+    x: Bytes[4] = msg.data[4]
+    return x
+"""
+    contract = get_contract(code)
+    assert contract.foo().hex() == "c2985578"  # fn sig
+
+
+def test_msg_data_call_args(get_contract):
+    code = """
+@external
+def foo(bar: uint256) -> (Bytes[4], uint256):
+    data: Bytes[36] = msg.data[36]
+    return (slice(data, 0, 4), convert(slice(data, 4, 32), uint256))
+"""
+    contract = get_contract(code)
+    a, b = contract.foo(42)
+
+    assert a.hex() == "2fbebd38"  # fn sig
+    assert b == 42
+
+
+def test_msg_data_bounds_check(get_contract):
+    code = """
+@external
+def foo(bar: uint256) -> Bytes[4]:
+    data: Bytes[4] = msg.data[4]  # msg.data.size > 4 (max_len)
+    return data
+"""
+    contract = get_contract(code)
+
+    with pytest.raises(TransactionFailed):
+        contract.foo(42)
```

Comment:
```suggestion
@external
def foo(bar: uint256) -> Bytes[4]:
    data: Bytes[4] = slice(msg.data, 0, 4)  # msg.data.size > 4 (max_len)
    return data
"""
    contract = get_contract(code)

    contract.foo(42).hex() == "00000000"  # EVM is big endian
```

Reply:
Add a runtime failure case like this:

```python
def foo() -> Bytes[4]:
    return slice(msg.data, 0, 4)  # Throws because len(msg.data) == 0
```